src/control/server/storage/scm/firmware.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/daos-stack/daos/src/control/server/storage"
+)
+
+type (
+	// firmwareProvider is an embedded structure that enables a Provider to
+	// forward firmware requests to a privileged binary if firmware management
+	// is enabled in the build.
+	firmwareProvider struct {
+		fwFwd *FirmwareForwarder
+	}
+
+	// FirmwareQueryRequest defines the parameters for a firmware query.
+	FirmwareQueryRequest struct {
+		pbin.ForwardableRequest
+		Devices []string // requested device UIDs, empty for all
+	}
+
+	// ModuleFirmware represents the results of a firmware query for a specific
+	// SCM module.
+	ModuleFirmware struct {
+		Module storage.ScmModule
+		Info   *storage.ScmFirmwareInfo
+		Error  string
+	}
+
+	// FirmwareQueryResponse contains the results of a successful firmware query.
+	FirmwareQueryResponse struct {
+		Results []ModuleFirmware
+	}
+
+	// FirmwareUpdateRequest defines the parameters for a firmware update.
+	FirmwareUpdateRequest struct {
+		pbin.ForwardableRequest
+		Devices      []string // requested device UIDs, empty for all
+		FirmwarePath string   // location of the firmware binary
+	}
+
+	// ModuleFirmwareUpdateResult represents the result of a firmware update for
+	// a specific SCM module.
+	ModuleFirmwareUpdateResult struct {
+		Module storage.ScmModule
+		Error  string
+	}
+
+	// FirmwareUpdateResponse contains the results of the firmware update.
+	FirmwareUpdateResponse struct {
+		Results []ModuleFirmwareUpdateResult
+	}
+)
+
+// setupFirmwareProvider initializes the provider's firmware capabilities.
+func (p *Provider) setupFirmwareProvider(log logging.Logger) {
+	p.fwFwd = NewFirmwareForwarder(log)
+}
+
+// QueryFirmware fetches the status of SCM device firmware.
+func (p *Provider) QueryFirmware(req FirmwareQueryRequest) (*FirmwareQueryResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Query(req)
+	}
+
+	modules, err := p.getRequestedModules(req.Devices)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &FirmwareQueryResponse{
+		Results: make([]ModuleFirmware, 0, len(modules)),
+	}
+	for _, mod := range modules {
+		fwInfo, err := p.backend.GetFirmwareStatus(mod.UID)
+		result := ModuleFirmware{
+			Module: *mod,
+			Info:   fwInfo,
+		}
+		if err != nil {
+			result.Error = err.Error()
+		}
+		resp.Results = append(resp.Results, result)
+	}
+
+	return resp, nil
+}
+
+// UpdateFirmware updates the SCM device firmware.
+func (p *Provider) UpdateFirmware(req FirmwareUpdateRequest) (*FirmwareUpdateResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Update(req)
+	}
+
+	if len(req.FirmwarePath) == 0 {
+		return nil, errors.New("missing path to firmware file")
+	}
+
+	modules, err := p.getRequestedModules(req.Devices)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(modules) == 0 {
+		return nil, errors.New("no SCM modules")
+	}
+
+	resp := &FirmwareUpdateResponse{
+		Results: make([]ModuleFirmwareUpdateResult, 0, len(modules)),
+	}
+	for _, mod := range modules {
```

Comment:
any plans for this to be done in parallel?

Reply:
We could certainly go down that road, though I'm not sure if there are limitations on parallelism in the kernel drivers. Requires a little further research.